book/optimization.qmd

```diff
@@ -772,12 +592,404 @@ It is therefore important to ensure that the performance of a tuned model is *al
+  geom_histogram(
+    bins = 15,
+    fill = viridis(1, begin = 0.5),
+    alpha = 0.8,
+    color = "black") +
+  theme_minimal()
+
+data = data.frame(cost = exp(cost))
+ggplot(data, aes(x = cost)) +
+  geom_histogram(
+    bins = 15,
+    fill = viridis(1, begin = 0.5),
+    alpha = 0.8,
+    color = "black") +
+  theme_minimal()
+```
+
+To add the `exp()` transformation to a hyperparameter, we pass `logscale = TRUE` to `to_tune()`.
+
+```{r optimization-053}
+learner = lrn("classif.svm",
+  cost  = to_tune(1e-5, 1e5, logscale = TRUE),
+  gamma = to_tune(1e-5, 1e5, logscale = TRUE),
+  kernel = "radial",
+  type = "C-classification"
+)
+
+instance = tune(
+  method = tnr("grid_search", resolution = 5, batch_size = 5),
+  task = tsk("sonar"),
+  learner = learner,
+  resampling = rsmp("cv", folds = 3),
+  measures = msr("classif.ce")
+)
+
+instance$result
+```
+
+The column `x_domain` contains the hyperparameter values after the transformation i.e. `exp(5.76)` and `exp(-5.76)`:
+
+```{r optimization-054}
+instance$result$x_domain
+```
+
+### Creating Search Spaces from Learners {#sec-search-space-token}
+
+We will now explain in more detail how a search space is created from a Learner.
+This is done by setting values of a `r ref("Learner")`'s `r ref("ParamSet")` to so-called `r ref("TuneToken")`s, constructed with a `r ref("to_tune")` call.
+This can be done in the same way that other hyperparameters are set to specific values.
+It can be understood as the hyperparameters being tagged for later tuning.
+The resulting `r ref("ParamSet")` used for tuning can be retrieved using the `$search_space()` method.
+
+```{r optimization-046}
+learner = lrn("classif.svm")
+learner$param_set$set_values(
+  kernel = "polynomial", # for example
+  degree = to_tune(lower = 1, upper = 3)
+)
+learner$param_set$search_space()
+```
+
+It is possible to omit `lower` here, because it can be inferred from the lower bound of the `degree` parameter itself.
+For other parameters, that are already bounded, it is possible to not give any bounds at all, because their ranges are already bounded.
+An example is the logical `shrinking` hyperparameter:
+
+```{r optimization-047}
+learner$param_set$set_values(shrinking = to_tune())
+learner$param_set$search_space()
+```
+
+A `r ref("TuneToken")` can also be constructed with a `r ref("Domain")` object, i.e. something constructed with a `p_<type>` call.
+This way it is possible to tune continuous parameters with discrete values, or to give transformations or dependencies.
+One could, for example, tune the `cost` as above on three given special values, and introduce a dependency of `shrinking` on it.
+Notice that a short form for `to_tune(<levels>)` is a short form of `to_tune(p_fct(<levels>))`.
+
+:::{.callout-note}
+When introducing the dependency, we need to use the `degree` value from *before* the implicit transformation, which is the name or `as.character()` of the respective value, here `"val2"`!
```

Comment:
```suggestion
When introducing the dependency, we need to use the `cost` value from *before* the implicit transformation. Concretely, we write the dependency as `cost == "val2"` and not `cost == 0.7`.
```

Reply:
It's still confusing but is made clearer with the example